Neos.ContentRepositoryRegistry/Classes/NodeSerializer.php

```diff
@@ -0,0 +1,59 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Neos\ContentRepositoryRegistry;
+
+use Neos\ContentRepository\Core\Projection\ContentGraph\Node;
+use Neos\ContentRepository\Core\Projection\ContentGraph\VisibilityConstraints;
+use Neos\ContentRepository\Core\SharedModel\Node\NodeIdentity;
+use Neos\Flow\Annotations as Flow;
+
+/**
+ * Utility to convert the {@see Node} to its {@see NodeIdentity} and reverse.
+ *
+ * @api
+ */
+#[Flow\Scope('singleton')]
+final readonly class NodeSerializer
+{
+    public function __construct(
+        private ContentRepositoryRegistry $contentRepositoryRegistry
+    ) {
+    }
+
+    public function denormalizeNodeFromIdentity(NodeIdentity $identity): Node
+    {
+        $contentRepository = $this->contentRepositoryRegistry->get($identity->contentRepositoryId);
+        $workspace = $contentRepository->getWorkspaceFinder()->findOneByName($identity->workspaceName);
+        if (!$workspace) {
+            throw new \RuntimeException(sprintf('Workspace could not be found while deserializing node NodeIdentity<%s>.', json_encode($identity, JSON_PARTIAL_OUTPUT_ON_ERROR)), 1707757488);
```

Comment:
We mix the terms "(de)serialize" and "(de)normalize" but actually neither are correct.. In its current form it's rather a converter mixed with a repository..

The name `NodeSerializer` came from the idea to represent nodes in URLs etc.

Reply:
so

```php
public function findNodeByIdentity(NodeIdentity $identity): Node;

public function convertNodeToIdentity(Node $node): NodeIdentity;
```

?

Reply:
It's more about the `NodeSerializer` (and I'm not stumbling upon the naming alone, but it's usually a sign of a flawed abstraction if that doesn't fit).
Let's discuss it in our weekly later